### PR TITLE
[data] Avoid under-parallelization regressions and add better testing for parallelism detection

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

In the previous PR https://github.com/ray-project/ray/pull/25883, a subtle regression was introduced in the case where data sizes blow up significantly.

For example, suppose you're reading jpeg-image files from a Dataset, which increase in size substantially on decompression. On a small-core cluster (e.g., 4 cores), you end up with 4-8 blocks of ~200MiB each when reading a 1GiB dataset. This can blow up to OOM the node when decompressed (e.g., 25x size increase).

Previously the heuristic to use parallelism=200 avoids this small-node problem. This PR avoids this issue by (1) raising the min parallelism back to 200. As an optimization, we also introduce the min block size threshold, which allows using fewer blocks if the data size is really small (<100KiB per block).

### CODE REVIEW

File: python/ray/data/_internal/util.py [hunk 1]
```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
...
+        )
+        logger.debug(
+            f"Autodetected parallelism={parallelism} based on "
+            f"estimated_available_cpus={avail_cpus} and "
+            f"estimated_data_size={mem_size}."
+        )
+    return parallelism, min_safe_parallelism
+
+
+def _estimate_avail_cpus(cur_pg: Optional["PlacementGroup"]) -> int:
```
Comment (ericl) [inline python/ray/data/_internal/util.py:120]: Moved verbatim.

File: python/ray/data/_internal/util.py [hunk 2]
```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
```
Comment (stephanie-wang) [inline python/ray/data/_internal/util.py:67]: Nit, I think it is more understandable to pass all dependency vars in directly instead of grabbing them from the context.

File: python/ray/data/_internal/util.py [hunk 3]
```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
...
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on three heuristic functions:
```
Comment (stephanie-wang) [inline python/ray/data/_internal/util.py]: Shouldn't ctx.min_parallelism also be included in this list?

Also, I found the heuristic description a little bit hard to follow because I wasn't really sure about priority between the different rules. Might help to also say which cases we cannot avoid breaking a rule. Also to say which ones lower-bound vs upper-bound the parallelism.

Comment (ericl) [reply python/ray/data/_internal/util.py]: I updated this to include the min_parallelism and state the precedence in order.

File: python/ray/data/_internal/util.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/_internal/util.py:67]: Updated to pass the context explicitly (I think it's too much to pass it all manually).

File: python/ray/data/_internal/util.py [hunk 4]
```diff
@@ -51,3 +58,97 @@ def _check_pyarrow_version():
...
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    ctx: DatasetContext,
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on the following heuristics (highest to lowest
```
Comment (stephanie-wang) [inline python/ray/data/_internal/util.py:72]: Maybe nit, but I feel like it's odd to say that this is highest to lowest precedence. The way I read this is, for example, that we would prefer to use num available CPUs even if it disrespects the max block size.

File: python/ray/data/_internal/util.py [hunk 5]
```diff
@@ -51,3 +58,97 @@ def _check_pyarrow_version():
...
+    Returns:
+        Tuple of detected parallelism (only if -1 was specified), and the min safe
+        parallelism (which can be used to generate warnings about large blocks).
+    """
+    min_safe_parallelism = 1
+    if reader:
+        mem_size = reader.estimate_inmemory_data_size()
+        if mem_size is not None:
+            min_safe_parallelism = max(1, int(mem_size / ctx.target_max_block_size))
+            max_reasonable_parallelism = max(
```
Comment (jiaodong) [inline python/ray/data/_internal/util.py:99]: are these max() needed since the only case it applies is when our target max / min block size is greater than mem_size, which seems we should just throw ?

File: python/ray/data/_internal/util.py [hunk 6]
```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
...
+    else:
+        mem_size = None
+        max_reasonable_parallelism = 9999999
+    if parallelism < 0:
+        if parallelism != -1:
+            raise ValueError("`parallelism` must either be -1 or a positive integer.")
+        # Start with 2x the number of cores as a baseline, with a min floor.
+        avail_cpus = avail_cpus or _estimate_avail_cpus(cur_pg)
+        parallelism = max(
+            min(ctx.min_parallelism, max_reasonable_parallelism),
```
[the remark this answers is not in the document]
Comment (jiaodong) [reply python/ray/data/_internal/util.py:111]: tbh i know this function is developed as-is today probably due to many legacy reasons but i found this function very critical for perf but also quite hard to read between all the min/max and nested min/max .. It looks high ROI to me to refactor this function completely into more modular functions, each covered by its own test so this decision making is clear for everyone else as a straight set of function calls for each rule, what do you think ?

File: python/ray/data/_internal/util.py (hunk 4 above)
Comment (ericl) [reply python/ray/data/_internal/util.py]: Yeah, I changed this to just specify an order, hopefully it's better now.

File: python/ray/data/_internal/util.py (hunk 5 above)
Comment (ericl) [reply python/ray/data/_internal/util.py:98]: We should never throw; those are signals to adjust the parallelism to meet the requirements.

File: python/ray/data/_internal/util.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/util.py:110]: Nack. We have a test added that covers this function in all the edge scenarios... I think that is more than enough. It's not useful to test them in smaller units since it's the interactions that are important.

File: python/ray/data/context.py [hunk 7]
```diff
@@ -11,8 +11,14 @@
 _context_lock = threading.Lock()
 
 # The max target block size in bytes for reads and transformations.
+# We choose 512MiB as 8x less than the typical memory:core ratio of 4:1.
 DEFAULT_TARGET_MAX_BLOCK_SIZE = 512 * 1024 * 1024
 
+# Datasets will avoid creating blocks smaller than this size in bytes on read.
+# This takes precedence over MIN_SAFE_PARALLELISM.
+# We choose 100KiB to be consistent with the Ray object inline threshold (100KiB).
+DEFAULT_TARGET_MIN_BLOCK_SIZE = 100 * 1024
```
Comment (clarkzinzow) [inline python/ray/data/context.py:20]: 100 KiB seems too small for Ray's task and general object overhead, are we sure we don't want a minimum in the range of 1-128 MiB?

File: python/ray/data/tests/test_auto_parallelism.py [hunk 8]
```diff
@@ -1,10 +1,106 @@
...
+        ctx=DatasetContext.get_current(),
+        reader=MockReader(test.data_size),
+        avail_cpus=test.avail_cpus,
+    )
+    assert result == test.expected_parallelism, (result, test)
+
+
+def test_autodetect_parallelism():
+    for test in TEST_CASES:
+        run_test(test)
```
Comment (clarkzinzow) [inline python/ray/data/tests/test_auto_parallelism.py:101]: Could you instead run these test cases via parametrizing the test, so there's one test run per test case? In addition to that being more idiomatic pytest test design, if a single test case fails, it will be a lot easier to rerun that specific test.

File: python/ray/data/_internal/util.py [hunk 9]
```diff
@@ -51,3 +58,96 @@ def _check_pyarrow_version():
...
+
+    This detects parallelism using on the following heuristics, applied in order:
+
+     1) We start with the default parallelism of 200.
+     2) Min block size. If the parallelism would make blocks smaller than this
+        threshold, the parallelism is reduced to avoid the overhead of tiny blocks.
+     3) Max block size. If the parallelism would make blocks larger than this
+        threshold, the parallelism is increased to avoid OOMS during processing.
+     4) Available CPUs. If the parallelism cannot make use of all the available
+        CPUs in the cluster, the parallelism is increased until it can.
```
Comment (clarkzinzow) [inline python/ray/data/_internal/util.py:80]: Great documentation!

File: python/ray/data/_internal/util.py [hunk 10]
```diff
@@ -51,3 +58,96 @@ def _check_pyarrow_version():
...
+        ctx: The current Dataset context to use for configs.
+        reader: The datasource reader, to be used for data size estimation.
+        avail_cpus: Override avail cpus detection (for testing only).
+
+    Returns:
+        Tuple of detected parallelism (only if -1 was specified), and the min safe
+        parallelism (which can be used to generate warnings about large blocks).
+    """
+    min_safe_parallelism = 1
+    max_reasonable_parallelism = 9999999
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/util.py:94]: I was a bit afraid this would cast the entire thing to float in some situations, so the large int seemed safer.

File: python/ray/data/context.py (hunk 7 above)
Comment (ericl) [reply python/ray/data/context.py:20]: I'd be ok with 1MiB, but note that this optimization is only reducing the overhead of tiny datasets (e.g., example datasets), shouldn't matter too much in practice.

File: python/ray/data/tests/test_auto_parallelism.py (hunk 8 above)
Comment (ericl) [reply python/ray/data/tests/test_auto_parallelism.py:101]: Hmm, it seems kind of sketchy to parameterize with an TestCase object, it's not clear to me pytest param is good for table-driven tests like this. I also think this way is more straightforward / simpler if you're not a pytest expert.

File: python/ray/data/_internal/util.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/_internal/util.py:94]: Hmm `max`/`min` don't cast their outputs, so I don't think that we need to worry about that, and any future refactor that produces a float would most likely be returning infinity; such a bug should be caught by our tests.

If you're still concerned about a float sneaking in somewhere, IMO we should use `sys.maxsize` rather than this "maybe large enough" number. https://docs.python.org/3/library/sys.html#sys.maxsize

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/_internal/util.py:94]: sys.maxsize it is then

File: python/ray/data/tests/test_auto_parallelism.py [hunk 11]
```diff
@@ -1,10 +1,106 @@
...
 
+@dataclass
+class TestCase:
+    avail_cpus: int
+    data_size: int
+    expected_parallelism: int
+
+
+MB = 1024 * 1024
+GB = 1024 * MB
```
Comment (clarkzinzow) [inline python/ray/data/tests/test_auto_parallelism.py:18]: Nit: These are mebibytes and gibibytes, not megabytes and gigabytes.
```suggestion
MiB = 1024 * 1024
GiB = 1024 * MiB
```

File: python/ray/data/tests/test_auto_parallelism.py (hunk 8 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_auto_parallelism.py:101]: >it's not clear to me pytest param is good for table-driven tests like this

It's definitely built for patterns exactly like this:
```python
@pytest.mark.parametrize(
    "avail_cpus,data_size,expected",
    [
        (4, 1024, 8),
        (4, 1 * MB, 10),
        (4, 2 * MB, 20),
    ],
)
def test_autodetect_parallelism(avail_cpus, data_size, expected):
    print("TESTING", test)

    class MockReader:
        def __init__(self, size):
            self.size = size

        def estimate_inmemory_data_size(self):
            return self.size

    result, _ = _autodetect_parallelism(
        parallelism=-1,
        cur_pg=None,
        ctx=DatasetContext.get_current(),
        reader=MockReader(data_size),
        avail_cpus=avail_cpus,
    )
    assert result == expected, (result, test)
```

>I also think this way is more straightforward / simpler if you're not a pytest expert.

I disagree a bit there, this is one of the core features of pytest, and is a big dev UX improvement. pytest is our core testing library for Python code so I think asking devs to use this basic pytest feature when implementing such a test case pattern is reasonable.

IMO we should really be trying to move our unit test suites towards a consistent arrange-act-assert pattern for each individual test instead of having these big mono-tests, and this is the idiomatic pytest way to run multiple test cases through a single arrange-act-assert test.

File: python/ray/data/context.py (hunk 7 above)
Comment (clarkzinzow) [reply python/ray/data/context.py:20]: Ah, since any dataset larger than 20 MiB will be hitting the 200 default parallelism ceiling. I think that 1 MiB would be better, that should mitigate an odd-seeming number of read tasks being spanned for e.g. reading a 20 MiB dataset with 200 files on an 8 CPU core machine launching 200 read tasks.

Comment (ericl) [reply python/ray/data/context.py:20]: Alright, changed to 1MiB.

File: python/ray/data/tests/test_auto_parallelism.py (hunk 8 above)
Comment (ericl) [reply python/ray/data/tests/test_auto_parallelism.py:101]: I tried this out, but it's considerably less readable. When scanning the table, you just can't really tell what's going on since there aren't any kwargs for the test table.

If you can generate pytest cases given dataclasses, that would be a good alternative that wouldn't sacrifice the readability.

Comment (ericl) [reply python/ray/data/tests/test_auto_parallelism.py]: Update: turns out astuple() is pretty convenient, changed to use that.

**Review by stephanie-wang — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]
LGTM!

**Review by jianoaix — APPROVED** [review]
Looks good.
